core/src/main/java/com/linecorp/armeria/client/endpoint/WeightRampingUpStrategy.java

```diff
@@ -184,53 +170,61 @@ WeightedRandomDistributionEndpointSelector endpointSelector() {
-                }
 
-                // Check whether we can ramp up with the next scheduled endpointsRampingUpEntry.
-                if (shouldRampUpWithNextScheduledEntry()) {
-                    // unhandledNewEndpoints will be ramped up when updateWeightAndStep() is executed.
-                    unhandledNewEndpoints = newlyAddedEndpoints;
-                    return;
-                }
+            // clean up existing entries
+            for (EndpointsRampingUpEntry entry : rampingUpWindowsMap.values()) {
+                entry.endpointAndSteps().clear();
             }
+            endpointsFinishedRampingUp.clear();
+
+            // We add the new endpoints from this point
+            final Object2LongOpenHashMap<Endpoint> newCreatedTimestamps = new Object2LongOpenHashMap<>();
+            for (Endpoint endpoint : newEndpoints) {
+                // Set the cached created timestamps for the next iteration
+                final long createTimestamp = computeCreateTimestamp(endpoint);
+                newCreatedTimestamps.put(endpoint, createTimestamp);
+
+                // check if the endpoint is already finished ramping up
+                final int step = numStep(rampingUpIntervalNanos, ticker, createTimestamp);
+                if (step >= totalSteps) {
+                    endpointsFinishedRampingUp.add(endpoint);
+                    continue;
+                }
 
-            if (newlyAddedEndpoints.isEmpty()) {
-                // newlyAddedEndpoints is empty which means that endpointsFinishedRampingUp are changed.
-                // So rebuild the endpoint selector.
-                buildEndpointSelector();
-                return;
-            }
+                // Create a EndpointsRampingUpEntry if there isn't one already
+                final long window = windowIndex(createTimestamp);
+                if (!rampingUpWindowsMap.containsKey(window)) {
+                    // align the schedule to execute at the start of each window
+                    final long initialDelayNanos = initialDelayNanos(window);
+                    final ScheduledFuture<?> scheduledFuture = executor.scheduleAtFixedRate(
+                            () -> updateWeightAndStep(window), initialDelayNanos,
+                            rampingUpIntervalNanos, TimeUnit.NANOSECONDS);
+                    final EndpointsRampingUpEntry entry = new EndpointsRampingUpEntry(
+                            new HashSet<>(), scheduledFuture, ticker, rampingUpIntervalNanos);
+                    rampingUpWindowsMap.put(window, entry);
+                }
+                final EndpointsRampingUpEntry rampingUpEntry = rampingUpWindowsMap.get(window);
 
-            updateWeightAndStep(newlyAddedEndpoints);
+                final EndpointAndStep endpointAndStep = new EndpointAndStep(endpoint, step);
+                final int currentWeight = computeWeight(endpoint, step);
+                endpointAndStep.currentWeight(currentWeight);
```

Comment:
At first, I didn't want to do this because `EndpointAndStep` is used in tests extensively and I thought it was important that arbitrary weights could be set in tests.
However, giving a second try seems like tests pass so sounds good to me 👍